### ISSUE TITLE
Deadlock during anti-entropy sync because yamux is blocked on a read...

### ISSUE THREAD

**Issue by sean-** [hashicorp/consul#2107]
Agent `localState` is hung.  yamux should have a timeout on its IOs to allow for automatic recovery.

```
$ curl -v 'http://172.16.139.148:8500/v1/agent/services?pretty'
* Hostname was NOT found in DNS cache
*   Trying 172.16.139.148...
* Connected to 172.16.139.148 (172.16.139.148) port 8500 (#0)
> GET /v1/agent/services?pretty HTTP/1.1
> User-Agent: curl/7.35.0
> Host: 172.16.139.148:8500
> Accept: */*
>
[infinite hang]
```

Attached is the backtrace.  The important two goroutines are:

```
goroutine 25 [select, 88 minutes]:
github.com/hashicorp/consul/vendor/github.com/hashicorp/yamux.(*Stream).Read(0xc8200ba410, 0xc82022b000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/yamux/stream.go:125 +0x3f1
bufio.(*Reader).fill(0xc8203aa8a0)
        /opt/go/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).ReadByte(0xc8203aa8a0, 0xc820547f20, 0x0, 0x0)
        /opt/go/src/bufio/bufio.go:229 +0x7a
github.com/hashicorp/consul/vendor/github.com/hashicorp/go-msgpack/codec.(*ioDecReader).readn1(0xc82032a270, 0x7fd51c2fd228)
        /opt/gopath/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/go-msgpack/codec/decode.go:90 +0x42
github.com/hashicorp/consul/vendor/github.com/hashicorp/go-msgpack/codec.(*msgpackDecDriver).initReadNext(0xc820547ee0)
        /opt/gopath/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/go-msgpack/codec/msgpack.go:540 +0x41
github.com/hashicorp/consul/vendor/github.com/hashicorp/go-msgpack/codec.(*Decoder).decode(0xc8203aa900, 0xb28fa0, 0xc82053b2c0)
        /opt/gopath/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/go-msgpack/codec/decode.go:635 +0x69
github.com/hashicorp/consul/vendor/github.com/hashicorp/go-msgpack/codec.(*Decoder).Decode(0xc8203aa900, 0xb28fa0, 0xc82053b2c0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/go-msgpack/codec/decode.go:630 +0x74
github.com/hashicorp/consul/vendor/github.com/hashicorp/net-rpc-msgpackrpc.(*MsgpackCodec).read(0xc8203810c0, 0xb28fa0, 0xc82053b2c0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/net-rpc-msgpackrpc/codec.go:121 +0xd1
github.com/hashicorp/consul/vendor/github.com/hashicorp/net-rpc-msgpackrpc.(*MsgpackCodec).ReadResponseHeader(0xc8203810c0, 0xc82053b2c0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/net-rpc-msgpackrpc/codec.go:74 +0x4d
github.com/hashicorp/consul/vendor/github.com/hashicorp/net-rpc-msgpackrpc.CallWithCodec(0x7fd51c30d040, 0xc8203810c0, 0xe93980, 0x12, 0xd2d700, 0xc8203b8500, 0xb32000, 0x130d298, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/consul/vendor/github.com/hashicorp/net-rpc-msgpackrpc/client.go:29 +0x197
github.com/hashicorp/consul/consul.(*ConnPool).RPC(0xc82005ac30, 0xe01698, 0x3, 0x7fd51c2fce00, 0xc8204ba2a0, 0x2, 0xe93980, 0x12, 0xd2d700, 0xc8203b8500, ...)
        /opt/gopath/src/github.com/hashicorp/consul/consul/pool.go:396 +0x221
github.com/hashicorp/consul/consul.(*Server).forwardLeader(0xc82018a480, 0xe93980, 0x12, 0xd2d700, 0xc8203b8500, 0xb32000, 0x130d298, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/consul/consul/rpc.go:229 +0x211
github.com/hashicorp/consul/consul.(*Server).forward(0xc82018a480, 0xe93980, 0x12, 0x7fd51c30d0a8, 0xc8203b8500, 0xd2d700, 0xc8203b8500, 0xb32000, 0x130d298, 0xc800000000, ...)
        /opt/gopath/src/github.com/hashicorp/consul/consul/rpc.go:206 +0x19d
github.com/hashicorp/consul/consul.(*Catalog).Deregister(0xc820028108, 0xc8203b8500, 0x130d298, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/consul/consul/catalog_endpoint.go:78 +0xe9
reflect.Value.call(0xc59e60, 0xd7e118, 0x13, 0xdfcd20, 0x4, 0xc820237880, 0x3, 0x3, 0x0, 0x0, ...)
        /opt/go/src/reflect/value.go:435 +0x120d
reflect.Value.Call(0xc59e60, 0xd7e118, 0x13, 0xc820237880, 0x3, 0x3, 0x0, 0x0, 0x0)
        /opt/go/src/reflect/value.go:303 +0xb1
net/rpc.(*service).call(0xc820186240, 0xc8201861c0, 0xc8205527c0, 0xc820018600, 0xc82052ac40, 0xd2d700, 0xc8203b8500, 0x16, 0xb32000, 0x130d298, ...)
        /opt/go/src/net/rpc/server.go:383 +0x1c2
net/rpc.(*Server).ServeRequest(0xc8201861c0, 0x7fd51c301ad0, 0xc820545ac0, 0x0, 0x0)
        /opt/go/src/net/rpc/server.go:498 +0x294
github.com/hashicorp/consul/consul.(*Server).RPC(0xc82018a480, 0xe93980, 0x12, 0xd2d700, 0xc8203b84b0, 0xb32000, 0x130d298, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/consul/consul/server.go:716 +0x109
github.com/hashicorp/consul/command/agent.(*localState).deleteService(0xc82018a1b0, 0xc82054aa21, 0x1d, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/consul/command/agent/local.go:540 +0x2c2
github.com/hashicorp/consul/command/agent.(*localState).syncChanges(0xc82018a1b0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/consul/command/agent/local.go:481 +0x17b
github.com/hashicorp/consul/command/agent.(*localState).antiEntropy(0xc82018a1b0, 0xc82000a780)
        /opt/gopath/src/github.com/hashicorp/consul/command/agent/local.go:339 +0x208
created by github.com/hashicorp/consul/command/agent.(*Agent).StartSync
        /opt/gopath/src/github.com/hashicorp/consul/command/agent/agent.go:592 +0x4f
```

and:

```
goroutine 9129 [semacquire, 80 minutes]:
sync.runtime_Semacquire(0xc82018a1c0)
        /opt/go/src/runtime/sema.go:47 +0x26
sync.(*RWMutex).RLock(0xc82018a1b4)
        /opt/go/src/sync/rwmutex.go:37 +0x58
github.com/hashicorp/consul/command/agent.(*localState).Services(0xc82018a1b0, 0x0)
        /opt/gopath/src/github.com/hashicorp/consul/command/agent/local.go:185 +0x81
github.com/hashicorp/consul/command/agent.(*HTTPServer).AgentServices(0xc82012c550, 0x7fd51c301d20, 0xc8200baf70, 0xc8203d96c0, 0x0, 0x0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/consul/command/agent/agent_endpoint.go:36 +0x48
github.com/hashicorp/consul/command/agent.(*HTTPServer).AgentServices-fm(0x7fd51c301d20, 0xc8200baf70, 0xc8203d96c0, 0x0, 0x0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/consul/command/agent/http.go:224 +0x5a
github.com/hashicorp/consul/command/agent.(*HTTPServer).wrap.func1(0x7fd51c301d20, 0xc8200baf70, 0xc8203d96c0)
        /opt/gopath/src/github.com/hashicorp/consul/command/agent/http.go:331 +0x4f4
net/http.HandlerFunc.ServeHTTP(0xc820137b40, 0x7fd51c301d20, 0xc8200baf70, 0xc8203d96c0)
        /opt/go/src/net/http/server.go:1618 +0x3a
net/http.(*ServeMux).ServeHTTP(0xc82012f6b0, 0x7fd51c301d20, 0xc8200baf70, 0xc8203d96c0)
        /opt/go/src/net/http/server.go:1910 +0x17d
net/http.serverHandler.ServeHTTP(0xc820138800, 0x7fd51c301d20, 0xc8200baf70, 0xc8203d96c0)
        /opt/go/src/net/http/server.go:2081 +0x19e
net/http.(*conn).serve(0xc820019c80)
        /opt/go/src/net/http/server.go:1472 +0xf2e
created by net/http.(*Server).Serve
        /opt/go/src/net/http/server.go:2137 +0x44e
```

or:

```
goroutine 269 [semacquire, 88 minutes]:
sync.runtime_Semacquire(0xc82018a1c0)
        /opt/go/src/runtime/sema.go:47 +0x26
sync.(*RWMutex).RLock(0xc82018a1b4)
        /opt/go/src/sync/rwmutex.go:37 +0x58
github.com/hashicorp/consul/command/agent.(*localState).Checks(0xc82018a1b0, 0x0)
        /opt/gopath/src/github.com/hashicorp/consul/command/agent/local.go:287 +0x81
github.com/hashicorp/consul/command/agent.(*Agent).RemoveService(0xc82018a180, 0xc82054aa21, 0x1d, 0x1, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/consul/command/agent/agent.go:837 +0x2dc
github.com/hashicorp/consul/command/agent.(*HTTPServer).AgentDeregisterService(0xc82012c550, 0x7fd51c301d20, 0xc82013c9c0, 0xc8203875e0, 0x0, 0x0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/consul/command/agent/agent_endpoint.go:296 +0x164
github.com/hashicorp/consul/command/agent.(*HTTPServer).AgentDeregisterService-fm(0x7fd51c301d20, 0xc82013c9c0, 0xc8203875e0, 0x0, 0x0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/consul/command/agent/http.go:238 +0x5a
github.com/hashicorp/consul/command/agent.(*HTTPServer).wrap.func1(0x7fd51c301d20, 0xc82013c9c0, 0xc8203875e0)
        /opt/gopath/src/github.com/hashicorp/consul/command/agent/http.go:331 +0x4f4
net/http.HandlerFunc.ServeHTTP(0xc820137e60, 0x7fd51c301d20, 0xc82013c9c0, 0xc8203875e0)
        /opt/go/src/net/http/server.go:1618 +0x3a
net/http.(*ServeMux).ServeHTTP(0xc82012f6b0, 0x7fd51c301d20, 0xc82013c9c0, 0xc8203875e0)
        /opt/go/src/net/http/server.go:1910 +0x17d
net/http.serverHandler.ServeHTTP(0xc820138800, 0x7fd51c301d20, 0xc82013c9c0, 0xc8203875e0)
        /opt/go/src/net/http/server.go:2081 +0x19e
net/http.(*conn).serve(0xc820250000)
        /opt/go/src/net/http/server.go:1472 +0xf2e
created by net/http.(*Server).Serve
        /opt/go/src/net/http/server.go:2137 +0x44e
```

[Hung_consul___.txt](https://github.com/hashicorp/consul/files/307694/Hung_consul___.txt)

**Comment by mfischer-zd** [comment]
Sean, we think we are running into this as well.  How can we determine whether we are similarly affected?  

Do you know when this issue was introduced?

**Comment by ubootda** [comment]
👍

**Comment by epssy** [comment]
:+1:

**Comment by slackpad** [comment]
The anti-entropy sync mechanism was completely reworked in 1.0.1, with a mind to decoupling the lock logic and making the overall design much more straightforward. Closing this out as I suspect that work should have addressed this. If this is seen post-1.0.1 we will open a new issue with a more relevant stack trace.
